### PR TITLE
Better support for RTL mode under GTK3

### PR DESCRIPTION
**Pull request by AliKet**
Replaced by #2194

I am doing thing PR to call for help!

### Motivation:

As i noticed that the **RTL** support is broken with some `wxControl`s (e.g. `wxGrid`)
when custom drawing is in play to draw the client area, and failing to de anything about it
with the current implementation, i tried with the mirrored DC approach and guess? it worked
like a charm! (more work to be done to get this ready, however).

### The problem: (now fixed by bb9c0fe and 281b736)

Even if everything looks right to me so far, I'm going crazy with one thing: **the srolling nightmare**!

To better understand what i am talking about, here is a patch to be applied to the **internat** sample:

1. Assuming you have chosen an RTL language (e.g. Arabic)
2. Try to scroll to the last column (by pressing LEFT key repeatedly)
3. The grid is correctly scrolled off to the right.
4. Now try to press the RIGHT key...
5. See the artifact caused by the scrolling!!

Sure, i'm missing something fundamental here about how scrolling should be setup to make things work as expected, so i need some guidance on how to achieve this with this PR.

TIA!

### The patch:
```
diff --git a/samples/internat/internat.cpp b/samples/internat/internat.cpp
index 7296519e3f..119842f432 100644
--- a/samples/internat/internat.cpp
+++ b/samples/internat/internat.cpp
@@ -30,6 +30,7 @@
 
 #include "wx/intl.h"
 #include "wx/file.h"
+#include "wx/grid.h"
 #include "wx/log.h"
 #include "wx/cmdline.h"
 
@@ -74,6 +75,8 @@ public:
     void OnTest3(wxCommandEvent& event);
     void OnTestMsgBox(wxCommandEvent& event);
 
+    void OnGrid(wxGridEvent& event);
+
     wxDECLARE_EVENT_TABLE();
 
     wxLocale& m_locale;
@@ -171,6 +174,8 @@ wxBEGIN_EVENT_TABLE(MyFrame, wxFrame)
     EVT_MENU(INTERNAT_TEST_2, MyFrame::OnTest2)
     EVT_MENU(INTERNAT_TEST_3, MyFrame::OnTest3)
     EVT_MENU(INTERNAT_TEST_MSGBOX, MyFrame::OnTestMsgBox)
+
+    EVT_GRID_SELECT_CELL(MyFrame::OnGrid)
 wxEND_EVENT_TABLE()
 
 wxIMPLEMENT_APP(MyApp);
@@ -337,11 +342,23 @@ MyFrame::MyFrame(wxLocale& locale)
     CreateStatusBar(1);
 
     // this demonstrates RTL layout mirroring for Arabic locales
+#if 0
     wxSizer *sizer = new wxBoxSizer(wxHORIZONTAL);
     sizer->Add(new wxStaticText(this, wxID_ANY, _("First")),
                 wxSizerFlags().Border());
     sizer->Add(new wxStaticText(this, wxID_ANY, _("Second")),
                 wxSizerFlags().Border());
+#endif
+    wxSizer *sizer = new wxBoxSizer(wxVERTICAL);
+
+    wxGrid* grid = new wxGrid(this, wxID_ANY);
+    grid->CreateGrid(5, 5);
+    grid->SetDefaultCellOverflow(false);
+    sizer->Add(grid, wxSizerFlags().Border());
+
+    grid->SetCellBackgroundColour(1, 1, *wxRED);
+
     SetSizer(sizer);
 }
 
@@ -541,3 +558,9 @@ void MyFrame::OnTestMsgBox(wxCommandEvent& WXUNUSED(event))
         wxMessageBox(_("Please report the details of your platform to us."));
     }
 }
+
+void MyFrame::OnGrid(wxGridEvent& event)
+{
+    event.Skip();
+    wxLogDebug("Cell(%d, %d)", event.GetRow(), event.GetCol());
+}
```

### CODE REVIEW

**Comment by AliKet** [comment]
Question to @vadz @paulcor 

Under **GTK3** what's the point of calling `wxPizza :: scroll()` to scroll the client area (i.e. `m_wxwindow`)  ?  isn't it supposed to have a native scrolling support for free (knowing that it's actualy added to a `GtkScrolledWindow`  ?

**Comment by AliKet** [comment]
@vadz  @paulcor 

Any objections/suggestions about the changes done so far on this PR ?

I'll ask for a full review when finishing this.

TIA!

**Comment by vadz** [comment]
Thanks for your work on this!

Unfortunately, I don't know anything about GTK RTL support, so I'm not really qualified to review this. If it fixes the problem, it should be applied, of course, but I wonder if we could somehow reduce the differences between

1. GTK 2 and GTK 3
2. GTK and the other ports

Notably the need for `#ifdef` in `wxScrollHelperBase::DoPrepareDC()` is really annoying, this is generic code, so if it can't be written to work in the same way with both wxGTK and the other ports, it means our API doesn't work in the same under all platforms. Can we somehow hide this?

File: src/gtk/window.cpp [hunk 1]
```diff
@@ -5077,23 +5083,23 @@ void wxWindowGTK::GTKSendPaintEvents(const GdkRegion* region)
 #endif
 {
 #ifdef __WXGTK3__
+    if ( GetLayoutDirection() == wxLayout_RightToLeft )
     {
```
Comment (paulcor) [inline src/gtk/window.cpp:5087]: I'm pretty sure setting the clip needs to stay.

File: src/gtk/window.cpp [hunk 2]
```diff
@@ -5077,23 +5083,23 @@ void wxWindowGTK::GTKSendPaintEvents(const GdkRegion* region)
...
     }
-    double x1, y1, x2, y2;
-    cairo_clip_extents(cr, &x1, &y1, &x2, &y2);
 
-    if (x1 >= x2 || y1 >= y2)
-        return;
+    GdkRectangle rect;
+    if ( !gdk_cairo_get_clip_rectangle(cr, &rect) )
+      return;
+
```
Comment (paulcor) [inline src/gtk/window.cpp:5097]: This is the same clip you just retrieved, no need to set it.

File: src/gtk/window.cpp [hunk 3]
```diff
@@ -5077,23 +5083,23 @@ void wxWindowGTK::GTKSendPaintEvents(const GdkRegion* region)
...
     {
-        cairo_region_t* region = gdk_window_get_clip_region(gtk_widget_get_window(m_wxwindow));
-        cairo_rectangle_int_t rect;
-        cairo_region_get_extents(region, &rect);
-        cairo_region_destroy(region);
-        cairo_rectangle(cr, rect.x, rect.y, rect.width, rect.height);
-        cairo_clip(cr);
+        const int width = gdk_window_get_width(GTKGetDrawingWindow());
+
+        cairo_translate(cr, width, 0);
```
Comment (paulcor) [inline src/gtk/window.cpp:5090]: I think this is going to mirror bitmaps and blits.

**Review by paulcor — COMMENTED** [review]
I've made a few initial comments, it will be a while before I can look at this thoroughly.

**Comment by paulcor** [comment]
I see now I should have put the comments **below** the code I was commenting on instead of above, so that's going to look a little confusing.

**Comment by paulcor** [comment]
Thinking about this a little more, I can't see how mirroring all drawing could be the right thing to do. What if you're drawing a map?

File: src/gtk/window.cpp (hunk 1 above)
Comment (AliKet) [reply src/gtk/window.cpp:5087]: I tried these samples with the clip removed (LTR/RTL) :
- dataview
- drawing
- griddemo
- listctrl
- treectrl
- widgets

can you elaborate on why the clip should be altered at all so we can add a comment explaining why it should be done like that ?

Comment (paulcor) [reply src/gtk/window.cpp:5087]: You can use `git blame` to identify the commit that added that code (bca7313499c11a6d7fecd2baa355ac09fd3ac83b in this case). Commit messages will often give you some context. I admit that code could probably use a comment. But in general, you should be very cautious about removing code.

**Comment by paulcor** [comment]
Some research indicates that mirroring is in fact how RTL is commonly done.

**Comment by AliKet** [comment]
Replaced by #2194

**Comment by AliKet** [comment]
> Thinking about this a little more, I can't see how mirroring all drawing could be the right thing to do. What if you're drawing a map?

As a solution to this problem (which _**fortunately**_ not so common in practice)
we already have the ability to set the layout direction explicitly for a particular window (i.e. by `SetLayoutDirection()`) and force it to be `wxLayout_LeftToRight`. (i tried it and it works for me)

**Pull request closed without merging**